File: app/Models/Reservation.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use App\Enums\ReservationStatus;
use Carbon\Carbon;

/**
 * @property int $id
 * @property int $table_id
 * @property int|null $user_id
 * @property string $nom_client
 * @property string $telephone
 * @property string|null $email
 * @property \Illuminate\Support\Carbon $date_reservation
 * @property \Illuminate\Support\Carbon $heure_debut
 * @property \Illuminate\Support\Carbon|null $heure_fin
 * @property int $duree
 * @property int $nombre_personnes
 * @property float $prix_total
 * @property float|null $acompte
 * @property ReservationStatus $statut
 * @property string|null $notes
 * @property \Illuminate\Support\Carbon|null $created_at
 * @property \Illuminate\Support\Carbon|null $updated_at
 * 
 * @property-read Table|null $table
 * @property-read User|null $user
 * @property-read string $statut_display
 */
class Reservation extends Model
{
    use HasFactory;

    protected $fillable = [
        'table_id',
        'user_id',
        'nom_client',
        'telephone',
        'email',
        'date_reservation',
        'heure_debut',
        'heure_fin',
        'duree',
        'nombre_personnes',
        'prix_total',
        'acompte',
        'statut',
        'notes',
    ];

    protected $casts = [
        'date_reservation' => 'date',
        'heure_debut' => 'datetime:H:i',
        'heure_fin' => 'datetime:H:i',
        'duree' => 'integer',
        'nombre_personnes' => 'integer',
        'prix_total' => 'decimal:2',
        'acompte' => 'decimal:2',
        'statut' => ReservationStatus::class,
    ];

    /**
     * Relation avec la table
     */
    public function table()
    {
        return $this->belongsTo(Table::class);
    }

    /**
     * Relation avec l'utilisateur (client)
     */
    public function user()
    {
        return $this->belongsTo(User::class);
    }

    /**
     * Vérifier si la réservation est confirmée
     */
    public function isConfirmee(): bool
    {
        return $this->statut === ReservationStatus::Confirmee;
    }

    /**
     * Confirmer la réservation
     */
    public function confirmer(): void
    {
        $this->statut = ReservationStatus::Confirmee;
        $this->save();
        
        // Marquer la table comme réservée
        $table = $this->getAttribute('table');
        if ($table) {
            $table->reserver();
        }
    }

    /**
     * Annuler la réservation
     */
    public function annuler(): void
    {
        $this->statut = ReservationStatus::Annulee;
        $this->save();
        
        // Libérer la table si elle était réservée pour cette réservation
        $table = $this->getAttribute('table');
        if ($table && $table->statut->value === 'reservee') {
            // Vérifier qu'il n'y a pas d'autres réservations actives pour cette table
            $hasOtherReservations = self::where('table_id', $this->table_id)
                ->where('id', '!=', $this->id)
                ->whereIn('statut', [
                    ReservationStatus::Attente->value,
                    ReservationStatus::Confirmee->value,
                    ReservationStatus::EnCours->value,
                ])
                ->whereDate('date_reservation', $this->date_reservation)
                ->exists();
            
            if (!$hasOtherReservations) {
                $table->liberer();
            }
        }
    }

    /**
     * Marquer la réservation comme en cours
     */
    public function marquerEnCours(): void
    {
        $this->statut = ReservationStatus::EnCours;
        $this->save();
        
        // Marquer la table comme occupée
        $table = $this->getAttribute('table');
        if ($table) {
            $table->occuper();
        }
    }

    /**
     * Terminer la réservation
     */
    public function terminer(): void
    {
        $this->statut = ReservationStatus::Terminee;
        $this->save();
        
        // Libérer la table
        $table = $this->getAttribute('table');
        if ($table) {
            $table->liberer();
        }
    }

    /**
     * Calculer le prix total selon la durée et le type de table
     */
    public function calculerPrix(): float
    {
        $table = $this->getAttribute('table');
        if (!$table) {
            return 0;
        }

        $prixParHeure = $table->prix_par_heure ?? 0;
        $prixFixe = $table->prix ?? 0;

        if ($prixParHeure > 0) {
            return $prixParHeure * $this->duree;
        }

        return $prixFixe;
    }

    /**
     * Accessor pour le statut affiché
     */
    public function getStatutDisplayAttribute(): string
    {
        return match($this->statut) {
            ReservationStatus::Attente => 'En attente',
            ReservationStatus::Confirmee => 'Confirmée',
            ReservationStatus::EnCours => 'En cours',
            ReservationStatus::Terminee => 'Terminée',
            ReservationStatus::Annulee => 'Annulée',
        };
    }

    /**
     * Scope pour les réservations à venir
     */
    public function scopeAVenir($query)
    {
        return $query->where('date_reservation', '>=', now()->toDateString())
                     ->whereIn('statut', [
                         ReservationStatus::Attente->value,
                         ReservationStatus::Confirmee->value,
                     ]);
    }

    /**
     * Scope pour les réservations du jour
     */
    public function scopeDuJour($query)
    {
        return $query->whereDate('date_reservation', today());
    }

    /**
     * Scope par statut
     */
    public function scopeOfStatut($query, $statut)
    {
        return $query->where('statut', $statut);
    }
}

File: resto-app/lib/models/reservation.dart
import 'table.dart';

enum ReservationStatus {
  attente,
  confirmee,
  enCours,
  terminee,
  annulee;

  String get displayName {
    switch (this) {
      case ReservationStatus.attente:
        return 'En attente';
      case ReservationStatus.confirmee:
        return 'Confirmée';
      case ReservationStatus.enCours:
        return 'En cours';
      case ReservationStatus.terminee:
        return 'Terminée';
      case ReservationStatus.annulee:
        return 'Annulée';
    }
  }

  static ReservationStatus fromString(String value) {
    switch (value) {
      case 'attente':
        return ReservationStatus.attente;
      case 'confirmee':
        return ReservationStatus.confirmee;
      case 'en_cours':
        return ReservationStatus.enCours;
      case 'terminee':
        return ReservationStatus.terminee;
      case 'annulee':
        return ReservationStatus.annulee;
      default:
        return ReservationStatus.attente;
    }
  }
}

class Reservation {
  final int id;
  final int? tableId;
  final Table? table;
  final int? userId;
  final String nomClient;
  final String telephone;
  final String? email;
  final DateTime dateReservation;
  final String heureDebut;
  final String? heureFin;
  final int duree;
  final int nombrePersonnes;
  final double prixTotal;
  final double? acompte;
  final ReservationStatus statut;
  final String? notes;
  final DateTime createdAt;
  final DateTime updatedAt;

  Reservation({
    required this.id,
    this.tableId,
    this.table,
    this.userId,
    required this.nomClient,
    required this.telephone,
    this.email,
    required this.dateReservation,
    required this.heureDebut,
    this.heureFin,
    required this.duree,
    required this.nombrePersonnes,
    required this.prixTotal,
    this.acompte,
    required this.statut,
    this.notes,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Reservation.fromJson(Map<String, dynamic> json) {
    // Helper pour convertir en int de manière sécurisée
    int parseInt(dynamic value) {
      if (value is int) return value;
      if (value is String) return int.tryParse(value) ?? 0;
      return 0;
    }

    // Helper pour convertir en double de manière sécurisée
    double parseDouble(dynamic value) {
      if (value is double) return value;
      if (value is int) return value.toDouble();
      if (value is String) return double.tryParse(value) ?? 0.0;
      return 0.0;
    }

    return Reservation(
      id: parseInt(json['id']),
      tableId: json['table_id'] != null ? parseInt(json['table_id']) : null,
      table: json['table'] != null
          ? Table.fromJson(json['table'] as Map<String, dynamic>)
          : null,
      userId: json['user_id'] != null ? parseInt(json['user_id']) : null,
      nomClient: json['nom_client'] as String? ?? '',
      telephone: json['telephone'] as String? ?? '',
      email: json['email'] as String?,
      dateReservation: DateTime.parse(
          json['date_reservation'] as String? ?? DateTime.now().toIso8601String()),
      heureDebut: json['heure_debut'] as String? ?? '00:00',
      heureFin: json['heure_fin'] as String?,
      duree: parseInt(json['duree']),
      nombrePersonnes: parseInt(json['nombre_personnes']),
      prixTotal: parseDouble(json['prix_total']),
      acompte: json['acompte'] != null ? parseDouble(json['acompte']) : null,
      statut: ReservationStatus.fromString(json['statut'] as String? ?? 'attente'),
      notes: json['notes'] as String?,
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'] as String)
          : DateTime.now(),
      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'] as String)
          : DateTime.now(),
    );
  }
}
